### ISSUE TITLE
gyp build fails with node v15.0.1 on linux

### ISSUE THREAD

**Issue by maxfliri** [kesla/node-snappy#198]
Hi,

I'm trying to install snappy with node v15.0.1 on linux, but I get an error from the the gyp build. Details and full build output are below.

node: v15.0.1
npm: 7.0.3
os: Ubuntu 18.04.5 LTS

```
$ npm i snappy
npm ERR! code 1
npm ERR! path /home/semaphore/node_modules/snappy
npm ERR! command failed
npm ERR! command sh -c prebuild-install || node-gyp rebuild
npm ERR! make: Entering directory '/home/semaphore/node_modules/snappy/build'
npm ERR!   CXX(target) Release/obj.target/snappy/deps/snappy/snappy-1.1.7/snappy-sinksource.o
npm ERR!   CXX(target) Release/obj.target/snappy/deps/snappy/snappy-1.1.7/snappy-stubs-internal.o
npm ERR!   CXX(target) Release/obj.target/snappy/deps/snappy/snappy-1.1.7/snappy.o
npm ERR!   AR(target) Release/obj.target/deps/snappy/snappy.a
npm ERR!   COPY Release/snappy.a
npm ERR!   CXX(target) Release/obj.target/binding/src/binding.o
npm ERR! binding.target.mk:118: recipe for target 'Release/obj.target/binding/src/binding.o' failed
npm ERR! make: Leaving directory '/home/semaphore/node_modules/snappy/build'
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp info using node-gyp@7.1.2
npm ERR! gyp info using node@15.0.1 | linux | x64
npm ERR! gyp info find Python using Python version 3.6.9 found at "/usr/bin/python3"
npm ERR! gyp http GET https://nodejs.org/download/release/v15.0.1/node-v15.0.1-headers.tar.gz
npm ERR! gyp http 200 https://nodejs.org/download/release/v15.0.1/node-v15.0.1-headers.tar.gz
npm ERR! gyp http GET https://nodejs.org/download/release/v15.0.1/SHASUMS256.txt
npm ERR! gyp http 200 https://nodejs.org/download/release/v15.0.1/SHASUMS256.txt
npm ERR! gyp info spawn /usr/bin/python3
npm ERR! gyp info spawn args [
npm ERR! gyp info spawn args   '/home/semaphore/.nvm/versions/node/v15.0.1/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
npm ERR! gyp info spawn args   'binding.gyp',
npm ERR! gyp info spawn args   '-f',
npm ERR! gyp info spawn args   'make',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/home/semaphore/node_modules/snappy/build/config.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/home/semaphore/.nvm/versions/node/v15.0.1/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/home/semaphore/.cache/node-gyp/15.0.1/include/node/common.gypi',
npm ERR! gyp info spawn args   '-Dlibrary=shared_library',
npm ERR! gyp info spawn args   '-Dvisibility=default',
npm ERR! gyp info spawn args   '-Dnode_root_dir=/home/semaphore/.cache/node-gyp/15.0.1',
npm ERR! gyp info spawn args   '-Dnode_gyp_dir=/home/semaphore/.nvm/versions/node/v15.0.1/lib/node_modules/npm/node_modules/node-gyp',
npm ERR! gyp info spawn args   '-Dnode_lib_file=/home/semaphore/.cache/node-gyp/15.0.1/<(target_arch)/node.lib',
npm ERR! gyp info spawn args   '-Dmodule_root_dir=/home/semaphore/node_modules/snappy',
npm ERR! gyp info spawn args   '-Dnode_engine=v8',
npm ERR! gyp info spawn args   '--depth=.',
npm ERR! gyp info spawn args   '--no-parallel',
npm ERR! gyp info spawn args   '--generator-output',
npm ERR! gyp info spawn args   'build',
npm ERR! gyp info spawn args   '-Goutput_dir=.'
npm ERR! gyp info spawn args ]
npm ERR! gyp info spawn make
npm ERR! gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
npm ERR! In file included from /home/semaphore/.cache/node-gyp/15.0.1/include/node/node.h:67:0,
npm ERR!                  from ../src/./binding.h:3,
npm ERR!                  from ../src/binding.cc:1:
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h: In member function ‘void v8::TracedReferenceBase<T>::SetSlotThreadSafe(T*)’:
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:912:22: error: expected type-specifier
npm ERR!      reinterpret_cast<std::atomic<T*>*>(&val_)->store(new_val,
npm ERR!                       ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:912:22: error: expected ‘>’
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:912:22: error: expected ‘(’
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:912:22: error: ‘atomic’ is not a member of ‘std’
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:912:35: error: expected primary-expression before ‘*’ token
npm ERR!      reinterpret_cast<std::atomic<T*>*>(&val_)->store(new_val,
npm ERR!                                    ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:912:36: error: expected primary-expression before ‘>’ token
npm ERR!      reinterpret_cast<std::atomic<T*>*>(&val_)->store(new_val,
npm ERR!                                     ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:912:38: error: expected primary-expression before ‘>’ token
npm ERR!      reinterpret_cast<std::atomic<T*>*>(&val_)->store(new_val,
npm ERR!                                       ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:913:54: error: ‘memory_order_relaxed’ is not a member of ‘std’
npm ERR!                                                       std::memory_order_relaxed);
npm ERR!                                                       ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:913:80: error: expected ‘)’ before ‘;’ token
npm ERR!                                                       std::memory_order_relaxed);
npm ERR!                                                                                 ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h: In member function ‘const T* v8::TracedReferenceBase<T>::GetSlotThreadSafe() const’:
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:920:29: error: expected type-specifier
npm ERR!      return reinterpret_cast<std::atomic<const T*> const*>(&val_)->load(
npm ERR!                              ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:920:29: error: expected ‘>’
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:920:29: error: expected ‘(’
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:920:29: error: ‘atomic’ is not a member of ‘std’
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:920:41: error: expected primary-expression before ‘const’
npm ERR!      return reinterpret_cast<std::atomic<const T*> const*>(&val_)->load(
npm ERR!                                          ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:920:41: error: expected ‘)’ before ‘const’
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:920:41: error: expected ‘;’ before ‘const’
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:920:49: error: expected unqualified-id before ‘>’ token
npm ERR!      return reinterpret_cast<std::atomic<const T*> const*>(&val_)->load(
npm ERR!                                                  ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:920:49: error: expected initializer before ‘>’ token
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h: At global scope:
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:1631:58: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!                                  Local<Value> export_value);
npm ERR!                                                           ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:3405:23: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    Local<Value> Name() const { return Description(); }
npm ERR!                        ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:5216:76: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!        ArrayBufferCreationMode mode = ArrayBufferCreationMode::kExternalized);
npm ERR!                                                                             ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:5263:21: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    bool IsExternal() const;
npm ERR!                      ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:5290:24: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    Contents Externalize();
npm ERR!                         ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:5300:70: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    void Externalize(const std::shared_ptr<BackingStore>& backing_store);
npm ERR!                                                                       ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:5311:24: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    Contents GetContents();
npm ERR!                         ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:5699:76: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!        ArrayBufferCreationMode mode = ArrayBufferCreationMode::kExternalized);
npm ERR!                                                                             ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:5748:76: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!        ArrayBufferCreationMode mode = ArrayBufferCreationMode::kExternalized);
npm ERR!                                                                             ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:5757:21: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    bool IsExternal() const;
npm ERR!                      ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:5773:24: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    Contents Externalize();
npm ERR!                         ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:5783:70: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    void Externalize(const std::shared_ptr<BackingStore>& backing_store);
npm ERR!                                                                       ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:5798:24: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    Contents GetContents();
npm ERR!                         ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:7143:28: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    size_t code_range_size() const { return code_range_size_ / kMB; }
npm ERR!                             ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:7145:46: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    void set_code_range_size(size_t limit_in_mb) {
npm ERR!                                               ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:7149:38: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    size_t max_semi_space_size_in_kb() const;
npm ERR!                                       ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:7151:56: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    void set_max_semi_space_size_in_kb(size_t limit_in_kb);
npm ERR!                                                         ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:7153:31: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    size_t max_old_space_size() const { return max_old_generation_size_ / kMB; }
npm ERR!                                ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:7155:49: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    void set_max_old_space_size(size_t limit_in_mb) {
npm ERR!                                                  ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:7159:31: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    size_t max_zone_pool_size() const { return max_zone_pool_size_; }
npm ERR!                                ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:7161:43: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    void set_max_zone_pool_size(size_t bytes) { max_zone_pool_size_ = bytes; }
npm ERR!                                            ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:8751:59: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!                                      MeasureMemoryMode mode);
npm ERR!                                                            ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:8814:36: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    Local<Context> GetEnteredContext();
npm ERR!                                     ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:9098:22: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    void RunMicrotasks() { PerformMicrotaskCheckpoint(); }
npm ERR!                       ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:9142:75: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    void AddMicrotasksCompletedCallback(MicrotasksCompletedCallback callback);
npm ERR!                                                                            ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:9150:78: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    void RemoveMicrotasksCompletedCallback(MicrotasksCompletedCallback callback);
npm ERR!                                                                               ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:9334:30: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    UnwindState GetUnwindState();
npm ERR!                               ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:9397:54: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!        AllowCodeGenerationFromStringsCallback callback);
npm ERR!                                                       ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:9757:75: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    static bool TryHandleSignal(int signal_number, void* info, void* context);
npm ERR!                                                                            ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:10713:55: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!                                  const void* stack_base);
npm ERR!                                                        ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/v8.h:10739:65: warning: ‘deprecated’ attribute directive ignored [-Wattributes]
npm ERR!    static bool PCIsInV8(const UnwindState& unwind_state, void* pc);
npm ERR!                                                                  ^
npm ERR! In file included from ../src/./binding.h:3:0,
npm ERR!                  from ../src/binding.cc:1:
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/node.h: In lambda function:
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/node.h:617:43: error: parameter packs not expanded with ‘...’:
npm ERR!              std::forward<EnvironmentArgs>(env_args)...);
npm ERR!                                            ^
npm ERR! /home/semaphore/.cache/node-gyp/15.0.1/include/node/node.h:617:43: note:         ‘env_args’
npm ERR! make: *** [Release/obj.target/binding/src/binding.o] Error 1
npm ERR! gyp ERR! build error
npm ERR! gyp ERR! stack Error: `make` failed with exit code: 2
npm ERR! gyp ERR! stack     at ChildProcess.onExit (/home/semaphore/.nvm/versions/node/v15.0.1/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:194:23)
npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:327:20)
npm ERR! gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:277:12)
npm ERR! gyp ERR! System Linux 4.15.0-122-generic
npm ERR! gyp ERR! command "/home/semaphore/.nvm/versions/node/v15.0.1/bin/node" "/home/semaphore/.nvm/versions/node/v15.0.1/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
npm ERR! gyp ERR! cwd /home/semaphore/node_modules/snappy
npm ERR! gyp ERR! node -v v15.0.1
npm ERR! gyp ERR! node-gyp -v v7.1.2
npm ERR! gyp ERR! not ok
```

**Comment by kunal15595** [comment]
does the latest version fix this issue?

**Comment by maxfliri** [comment]
Hi there,

I eventually found the cause for this, and it wasn't directly related to node-snappy. It turned out this was caused by an incorrect version of gcc/g++. The problem was occurring on our continuous integration agent, which has the rather old gcc version 4.8.5 installed by default. We switched to the latest gcc version available in that environment, 8.4.0, and the problem disappeared.

Apologies for not having updated the issue sooner. This can be closed now as far as I'm concerned. If anything, it would be good if you could document the requirements for compiling the native extensions, that would help troubleshooting this kind of issues.

**Comment by Brooooooklyn** [comment]
Try this one: https://github.com/Brooooooklyn/snappy, without node-gyp and postinstall scripts.
